### PR TITLE
Recover problem

### PR DESCRIPTION
**Pull request by Carnitas**

### CODE REVIEW

File: recover/recover.c [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+    fclose(card);
+}
+
+// Helper functions
+
+// Identify signature that starts a jpeg.
+bool jpg_signature_found(uint8_t buffer[BLOCK_SIZE])
+{
+    if (buffer[0] == 0xff && buffer[1] == 0xd8 && buffer[2] == 0xff &&
+        (buffer[3] >= 0xe0 && buffer[3] <= 0xef))
```
Comment (jgopel) [inline recover/recover.c:77]: **broader context:** What you've done here is perfectly correct, however it might be surprising to someone working in this domain. What you're really looking for in the 4th byte is that the upper nibble is e. Programmers in this space would probably express that with a bitmask, which would look something like this:

```suggestion
    static const BYTE_3_MASK = 0xf0;
    if (buffer[0] == 0xff && buffer[1] == 0xd8 && buffer[2] == 0xff &&
        (buffer[3] & BYTE_3_MASK) == 0xe0)
```

I don't think there's any reason to change what you have here - I just thought it would be worth noting in case you came across it in your future travels.

File: recover/recover.c [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        if (new_jpg_found)
+        {
+            new_jpg_found = false;
+        }
+
+        if (jpg_signature_found(buffer))
+        {
+            if (file != NULL)
+            {
+                fclose(file);
```
Comment (jgopel) [inline recover/recover.c:55]: **suggestion:** For your purposes here this is probably fine, but in real-world usage I would expect my developers to examine the return code of this function to ensure that the function succeeded. Same goes for all other usages of the file manipulation functions. These functions can and do fail in real-world usage - especially `fread()` and `fwrite()` with larger values. The OS is completely free to take whatever size you gave it, write part of that size, and then stop writing to go do something else.

File: recover/recover.c [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        return 2;
+    }
+
+    // Create a buffer to look for jpeg signatures
+    uint8_t buffer[512];
+    int counter = 0;
+    bool new_jpg_found = false;
+    FILE *file = NULL;
+
+    while (fread(buffer, sizeof(buffer), 1, card))
```
[the remark this answers is not in the document]
Comment (jgopel) [reply recover/recover.c]: Looks like you got to this before I pressed submit on the review - nice!

File: recover/recover.c [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+        if (jpg_signature_found(buffer))
+        {
+            if (file != NULL)
+            {
+                fclose(file);
+            }
+            file = create_jpg_file(counter);
+            counter++;
+        }
+        fwrite(buffer, 1, BLOCK_SIZE, file);
```
Comment (Carnitas) [inline recover/recover.c:52]: What would I assert here and how would I validate it?

File: recover/recover.c (hunk 1 above)
Comment (Carnitas) [reply recover/recover.c:64]: I was wondering how I could just look at the first four bytes and you can't do something like `buffer[3] == 0xe[0-f]`, which is too bad. I am going to add this suggestion so that I can reference this solution in the future. The comparison I used seems bulky.

File: recover/recover.c (hunk 4 above)
Comment (jgopel) [reply recover/recover.c:52]: The short answer is something like 
```suggestion
        const int bytes_written = fwrite(buffer, 1, BLOCK_SIZE, file);
        assert(bytes_written == BLOCK_SIZE);
```

The longer answer is that the man pages will tell you what the return value of all these functions is. For convenience, [here's the man page for `fwrite()`](https://man7.org/linux/man-pages/man3/fwrite.3p.html). If you look at the section on the return value, you'll see that it returns the number of elements that were written. The number of elements that you are trying to write is the 3rd parameter of `fwrite()`. Based on this, you know that everything was written if the return value is the same as `nitems`.

File: recover/recover.c [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+    int counter = 0;
+
+    FILE *file = create_jpg_file(counter);
+
+    while (fread(buffer, 1, BLOCK_SIZE, card) == BLOCK_SIZE)
+    {
+        if (jpg_signature_found(buffer))
+        {
+            if (file != NULL)
+            {
```
Comment (jgopel) [inline recover/recover.c:46]: I think `file` is never `NULL` now - right? Is there something in one of the functions you're calling that's mutating the value of that pointer?

File: recover/recover.c [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+    fclose(file);
+    fclose(card);
+}
+
+// Helper functions
+
+// Identify signature that starts a jpeg.
+bool jpg_signature_found(uint8_t buffer[BLOCK_SIZE])
+{
+    static const uint8_t BYTE_3_MASK = 0xe0;
```
Comment (jgopel) [inline recover/recover.c:63]: **issue:** This isn't correct anymore I don't think. A bitmask is used to select the bits that you want to keep. In this case, I think you want to keep the top 4 bits - as written you are only keeping the top 3 bits. What this leads to in this code is that `0xf*` is a valid 3rd byte - I think only `0xe*` is valid. When I was newer to writing bitmasks, something that really helped me was writing the binary notation for the mask - here you have `0b1110 0000` - I think you want `0b1111 0000`.

Comment (jgopel) [reply recover/recover.c:63]: Let me know if this comment doesn't make sense.

File: recover/recover.c (hunk 4 above)
Comment (Carnitas) [reply recover/recover.c:52]: `assert` is what I was looking for! Another `if` loop stops the buffer from updating unless I stick another `while` loop inside it.

File: recover/recover.c (hunk 6 above)
Comment (Carnitas) [reply recover/recover.c:63]: It doesn't. Per the constraints of the assignment under [background](https://cs50.harvard.edu/x/2024/psets/4/recover/), we should assume that the first three bytes of a jpg are the same each time. The first for bits of the 4th byte always begin `0xe` and then a value of `0-f`, so I think the original solution is sufficient (at least it builds and passes tests, for what that's worth).

Comment (Carnitas) [reply recover/recover.c:63]: Unless it would be more obvious to use `BYTE_4_MASK` but that might cause a different readability problem.

File: recover/recover.c (hunk 5 above)
Comment (Carnitas) [reply recover/recover.c:46]: After the first initialization, `file` should be `NULL` again. This format was provided by the lecture as a way to protect `fclose()` operations (should I should probably do again at line 56.

File: recover/recover.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (Carnitas) [reply recover/recover.c]: I tried an alternate solution based on lecture, `while (fread(buffer, 1, BLOCK_SIZE, card) == BLOCK_SIZE)`. What do you think? Or is this just restating the problem in a different way?

[the remark this answers is not in the document]
Comment (jgopel) [reply recover/recover.c]: Definitely a valid solution. I love that you're asking this question because it lets us talk a bit more about software design! As with most design questions, I think it's helpful to consider some trade offs here.

For me, performance is always front-of-mind. I would expect the runtime performance of both solutions to be similar up to the point that an error is encountered, in which case the `while` solution will probably produce its error slightly faster than the `assert` solution. That said (and you 100% have not learned this yet - this is getting into production concerns), there is a flag that I can pass (`-NDEBUG`) which will cause the compiler not to turn all `assert`s into no-ops. So if you are ok with not having the assertion in production (ie: testing with it enable to ensure your implementation is good, then disabling it for production on trust that your testing, data, etc have covered everything) the `assert` based solution has an edge.

I think it's also worth considering the ergonomics of the failure case. In the case that your `read` gets (eg) 42 bytes, this `while` solution will stop executing with no discernable difference from if it had reached the end of its data and encountered 0 bytes. I would ask myself what I would want to happen here - does it make sense for that termination to happen silently, or is this something that should take a different path so a human can notice it? In pset land it's hard to answer these questions, but in a real environment I think you'll find that those sorts of answers are usually pretty apparent.

File: recover/recover.c [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+bool jpg_signature_found(uint8_t buffer[BLOCK_SIZE]);
+FILE *create_jpg_file(int counter);
+
+int main(int argc, char *argv[])
+{
+    // Check for invalid usage
+    if (argc != 2)
+    {
+        printf("Usage: ./recover [filepath]\n");
+        return 1;
```
Comment (jgopel) [inline recover/recover.c:24]: **praise:** I love the different return codes for the different sorts of problems! If you wanted to supercharge this, you could do something like this:

```c
enum error_codes {
    // You need this first value explicitly because the default first value is 0
    error_incorrect_usage = 1,
    error_could_not_open_file,
    ...
};
```
```suggestion
        return error_incorrect_usage;
```

File: recover/recover.c (hunk 6 above)
Comment (jgopel) [reply recover/recover.c:74]: I think you're misunderstanding how bitwise AND works. I'll try and explain that a little better.

Let's consider the single bit case first:

| x | y | x & y |
| - | - | ------ |
| 0 | 0 | 0 |
| 0 | 1 | 0 |
| 1 | 0 | 0 |
| 1 | 1 | 1 |

This operation is applied for each bit. Some examples:
- `0b1010 & 0b1111 == 0b1010`
- `0b1010 & 0b0000 == 0b0000`
- `0b1010 & 0b1010 == 0b1010` (in fact, any value `&` with itself produces itself.
- `0b0101 & 0b1010 == 0b0000`

The case that I'm concerned about here is where you have `0xf1` (low byte is arbitrary), which is `0b1111 0001`. Your mask is `0xe0`, which is `0b1110 0000`. `0b1111 0001 & 0b1110 0000 == 0b1110 0000`, which is `0xe0`. So now `0xe0` is compared with `0xe0` and that returns true. I think that `0xf1` is not valid for the 4th (1 indexed) byte, which means that this logic has a bug.

File: recover/recover.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (Carnitas) [reply recover/recover.c]: I've heard of some production requirements insist on removing any test code, which could include code that we'd disable at the compiler, so I reckon we'd have to plan on scrubbing it. 

For now I'll mark it resolved  as I don't think we reach the complexity to make the edge cases apparent.

File: recover/recover.c (hunk 6 above)
Comment (Carnitas) [reply recover/recover.c:74]: I still don't fully get it, but I'll keep this conversation for the next opportunity to apply it.

File: recover/recover.c (hunk 7 above)
Comment (Carnitas) [reply recover/recover.c:24]: I don't understand the value of this approach. What is `return error_incorrect_usage` doing if the human-readable information is in the `printf` statement?

File: recover/recover.c (hunk 6 above)
Comment (Carnitas) [reply recover/recover.c:74]: Wait, I get it now. Let me take a shot at a fix.

File: recover/recover.c (hunk 7 above)
Comment (jgopel) [reply recover/recover.c:24]: The idea is to create a mapping between `$?` and the actual error without the person who encountered the error having to inspect the code. It also makes it easier for you to, eg, change the return codes or unify multiple different failure paths into a single return code.

Fundamentally it's removing a magic value and that has all sorts of benefits. You can see all sorts of discussion on the topic if you Google for it, but here are some that I liked after a quick search:
- https://pylint.readthedocs.io/en/latest/user_guide/messages/refactor/magic-value-comparison.html
- https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Res-magic
- https://www.reddit.com/r/learnprogramming/comments/14ze381/should_magic_numbers_be_avoided_at_all_cost_are/

Comment (Carnitas) [reply recover/recover.c]: Removing arbitrary error codes makes sense so I've replaced most errors with `enum` error codes. Is there a way to retain verbosity for the message at line 42, which indicates the file that cannot be opened? I don't think I want to set `infile` as a global just to be able to reference it in the `print_error` function below.

Comment (jgopel) [reply recover/recover.c]: It's definitely possible, but maybe a bit advanced for where you're at now. The approach I would take in C would be a variadic function which can take any number of parameters and then you can be clever about how you unpack those args. I think it's perfectly acceptable to keep the print statements inline.

Comment (jgopel) [reply recover/recover.c]: Oh, another approach you could take would be to return the error message as a string. Then you could print the error message from where the error was generated and append any additional information at that point.
